### DISCUSSION TITLE
CeleryExecutor is failing to launch tasks with redis error

### DISCUSSION THREAD

**Discussion by ybryan** [apache/airflow#36097] [Q&A] [answered]
Similar issue: https://github.com/apache/airflow/issues/33744#issuecomment-1790216854

I've deployed Airflow on GKE with the `apache/airflow:2.7.3-python3.8` image. When launching tasks as of today, I'm receiving this error message on `airflow-scheduler`:

```
{celery_executor.py:303} ERROR - Error sending Celery task: module 'redis' has no attribute 'client'
"Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/celery/executors/celery_executor_utils.py", line 212, in send_task_to_executor
    result = task_to_run.apply_async(args=[command], queue=queue)
  File "/home/airflow/.local/lib/python3.8/site-packages/celery/app/task.py", line 594, in apply_async
    return app.send_task(
  File "/home/airflow/.local/lib/python3.8/site-packages/celery/app/base.py", line 794, in send_task
    with self.producer_or_acquire(producer) as P:
  File "/home/airflow/.local/lib/python3.8/site-packages/celery/app/base.py", line 929, in producer_or_acquire
    producer, self.producer_pool.acquire, block=True,
  File "/home/airflow/.local/lib/python3.8/site-packages/celery/app/base.py", line 1344, in producer_pool
    return self.amqp.producer_pool
  File "/home/airflow/.local/lib/python3.8/site-packages/celery/app/amqp.py", line 590, in producer_pool
    self.app.connection_for_write()]
  File "/home/airflow/.local/lib/python3.8/site-packages/celery/app/base.py", line 826, in connection_for_write
    return self._connection(url or self.conf.broker_write_url, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/celery/app/base.py", line 877, in _connection
    return self.amqp.Connection(
  File "/home/airflow/.local/lib/python3.8/site-packages/kombu/connection.py", line 201, in __init__
    if not get_transport_cls(transport).can_parse_url:
  File "/home/airflow/.local/lib/python3.8/site-packages/kombu/transport/__init__.py", line 90, in get_transport_cls
    _transport_cache[transport] = resolve_transport(transport)
  File "/home/airflow/.local/lib/python3.8/site-packages/kombu/transport/__init__.py", line 75, in resolve_transport
    return symbol_by_name(transport)
  File "/home/airflow/.local/lib/python3.8/site-packages/kombu/utils/imports.py", line 59, in symbol_by_name
    module = imp(module_name, package=package, **kwargs)
  File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/airflow/.local/lib/python3.8/site-packages/kombu/transport/redis.py", line 281, in <module>
    class PrefixedRedisPipeline(GlobalKeyPrefixMixin, redis.client.Pipeline):
AttributeError: module 'redis' has no attribute 'client'"
```

I shelled into the `airflow-scheduler` pod:

```sh
airflow@airflow-scheduler-7bb8dc9dd7-27thc:/opt/airflow$ pip freeze | grep redis
apache-airflow-providers-redis==3.4.0
google-cloud-redis==2.13.2
redis==4.6.0
airflow@airflow-scheduler-7bb8dc9dd7-27thc:/opt/airflow$ pip freeze | grep celery
apache-airflow-providers-celery==3.4.1
celery==5.3.4

airflow@airflow-scheduler-7bb8dc9dd7-27thc:/opt/airflow$ python
Python 3.8.18 (default, Nov  1 2023, 14:44:54)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import redis
>>> from redis import client
>>>

```

**Comment by potiuk** [comment]
Wrong POD.  Mostly something on your celery workers causes redis to be upgraded. But what it is and how - impossible to say .

You need to check what your `celery` pods have.

**Reply by ybryan** [reply]
Both `airflow-worker` and `airflow-scheduler` are using the same image

**Reply by potiuk** [reply]
Likely you have something that makes your celery image to upgrade redis. Check it on running celery worker. No-one knows what you do there - until you check and see what happens there.

**Reply by potiuk** [reply, marked as answer]
Also, it's likely that you have mistake in your deployment and either old containers or ones that use different images. Look for typos and the like.

**Reply by ybryan** [reply]
This is probably the answer. I've updated our helm charts to a different image (`apache/airflow:2.7.3-python3.9`) and things are running now

**Comment by gemoftheworld** [comment]
Hello @potiuk @ybryan 

seeing same exact error in apache/airflow:2.7.3-python3.9 too. I was using apache/airflow:2.7.3-python3.8 as well and after seeing this discussion upgraded to apache/airflow:2.7.3-python3.9 and it worked for sometime, now I see it again. Any workaround for this ?

**Reply by potiuk** [reply]
I do not know the reason, so no, it would be great to know WHAT you have changed (you had to change something - it could not just **start** happening - some of your deployment, customisations etc. must have done something. If we know what that SOMETHING was, we might attempt to diagnose and fix it, but so far we have no idea.

**Reply by potiuk** [reply]
New software Installation, building image automatically, clearing caches, pulling images, etc. etc. - if we know what could have triggered it, then we can do something to check it - any information, explanation, details what happened would be helpful.

**Reply by gemoftheworld** [reply]
in Helm, I set workers.persistence.enabled=false since I read logs from elasticsearch but I see that there is a specific logs.persistence field as well. Then what's difference between these two persistency ? When I set workers.persistence.enabled=true earlier, it was using Statefulset and auto-provisioned PVC name was logs-airflow-worker-0. When I set  workers.persistence.enabled=false, it turned from Statefulset to Deployment. Don't know if it's related or not

**Reply by gemoftheworld** [reply]
Nevertheless, reverting back changes doesn't help I still get AttributeError: module 'redis' has no attribute 'client' although see this 

```
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import redis
>>> from redis import client
>>>
```

**Reply by potiuk** [reply]
My guess is that SOMEWHERE on your PYTHONPATH you have `redis` package or module and your changes made it appear somewhere (your DAG folders, plugin, or maybe in your other system libraries).  This is is standard behaviour of Python that if you use `generic` packages or modules, only one of them is imported - the first one.

There is a whole documentation about that in Airflow docs that you should read and understand how python module importing work and how you should avoid the overlaps of generic names: https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/modules_management.html  

If you also want a bit more precise instructions and warnings of what you should learn and avoid, here is an update to the documentaiton that you should read (it is scheduled to be released together with 2.8.2) https://github.com/apache/airflow/pull/36927/files 

Please read and digest those docs, and it would be great, when you find the cause you come back here and explain what went wrong in your deployment  so that others could also see it and avoid it.

**Comment by duncan-thomas-hpe** [comment]
For those still running into this, a detailed explanation of the cause can be found in https://github.com/apache/airflow/issues/41359
